File: TRPO/lab8/src/main/java/GPSCoordinateType.java
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.IOException;

public class GPSCoordinateType implements UserType {
    @Override
    public String typeName() {
        return "GPSCoordinate";
    }

    @Override
    public Object create() {
        return new GPSCoordinate();
    }

    @Override
    public Object clone(Object obj) {
        if (!(obj instanceof GPSCoordinate)) 
            throw new IllegalArgumentException("Invalid object type");
        GPSCoordinate original = (GPSCoordinate) obj;
        return new GPSCoordinate(original.getLatitude(), original.getLongitude(), original.getTimestamp());
    }

    @Override
    public Object readValue(InputStreamReader in) throws IOException {
        BufferedReader reader = new BufferedReader(in);
        return parseValue(reader.readLine());
    }

    @Override
    public Object parseValue(String ss) {
        return GPSCoordinate.fromString(ss);
    }

    @Override
    public Comparator getTypeComparator() {
        return new Comparator() {
            @Override
            public int compare(Object o1, Object o2) {
                GPSCoordinate g1 = (GPSCoordinate) o1;
                GPSCoordinate g2 = (GPSCoordinate) o2;
                return Double.compare(g1.getLatitude(), g2.getLatitude());
            }
        };
    }
}

File: TRPO/lab7/app/src/main/java/com/example/lab7/GPSCoordinateType.kt
package com.example.lab7

import java.io.InputStream
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

data class GPSCoordinate(val lat: Double, val lon: Double, val time: LocalDateTime) {
    override fun toString() = "%.6f;%.6f;%s".format(lat, lon, time.format(DateTimeFormatter.ISO_DATE_TIME))
}

class GPSCoordinateType : UserType {
    override fun typeName() = "GPSCoordinate"
    override fun create() = GPSCoordinate(0.0, 0.0, LocalDateTime.now())
    override fun clone(obj: Any) = (obj as GPSCoordinate).copy()
    override fun readValue(input: InputStream) = parseValue(input.bufferedReader().readLine())
    override fun parseValue(ss: String): Any {
        val parts = ss.split(";")
        return GPSCoordinate(
            parts[0].toDouble(),
            parts[1].toDouble(),
            LocalDateTime.parse(parts[2], DateTimeFormatter.ISO_DATE_TIME)
        )
    }
    override fun getComparator() = Comparator<Any> { o1, o2 ->
        (o1 as GPSCoordinate).time.compareTo((o2 as GPSCoordinate).time)
    }
}